Configurable keybinds

helloimhana123 opened the pull request:
Okay holy moly. Configurable keybinds! Yes!

Features:

* Assign an absolute key (based on us keyboard), these are prefixed with "US_". For example "Ctrl+US_F3".
* Assign any unicode character. These don't need a prefix. For example "Ctrl+ø". Please note that these characters are case sensitive.
* Assign up to 3 binds per action (this can easily be extended if necessary).
* Help text, escape menu text, window menu text (for MS windows), are all updated based on the current assignment.
* No more hidden keybinds, everything is visible in the help text.
* When using US_ kind of bind, mod keys must match exactly. This solves a lot of annoyance with having to check mods all the time. For example pressing "Ctrl+Shift+F3" will not trigger "Ctrl+US_F3" since the shift key was also pressed.
*  When using unicode bind the mod keys don't have to match exactly, but the ones specified to have to be pressed. For example to do "Ctrl+ø" I need to press "Ctrl+AltGr+d". This combination will be accepted since "Ctrl" was held. But it will still accept it even though AltGr was also pressed. (hard to explain this...)

This code isn't 100% finished. I'm going on vacation soon and I need to focus on that, so I couldn't quite finish this up. So I need your help to finish it!

What I've done to make sure this is working:

* I made a python script that makes sure every keybind is in use, and is initiated. Download: [scan_keybinds.zip](https://github.com/user-attachments/files/15890332/scan_keybinds.zip)
* I have gone over every help text and made sure this new version has every bind that was in the previous version.
* I have tested the keybinds on global, pattern edit, and palette editor pages.

What's missing:

* General testing.
* Biggest issue is I have not tested the sample list or instrument list pages.
* There are a few stragglers left of keybinds I didn't replace.
* Scrolling on numbers on sample list page is not working properly.
* On order list page I haven't tested the binds "Ctrl+US_B", "Ctrl+US_O", "Ctrl+Shift+US_O", since I can't figure out what they are supposed to be.
* I also don't know what "Alt+US_U" on instrument list is supposed to do so I removed it.

I hope the code is acceptable and that we can bring this to 100%!!! 

Fixes #24
Fixes #68 
Fixes #189 
Related to #389 
Will probably affect #468

helloimhana123 committed:
Bunch of string utils

helloimhana123 committed:
Initial keybinds implementation

helloimhana123 committed:
Global keybinds implementation

helloimhana123 committed:
Show correct keybinds in windows window menus

helloimhana123 committed:
Show correct key binding in page titles

helloimhana123 committed:
Show new keybinds in help text

helloimhana123 committed:
For titles, removed checking if shortcut is empty

helloimhana123 committed:
Add logging and errors to message log (ctrl-F11)

helloimhana123 committed:
Show current keybinds in escape menu

helloimhana123 committed:
Read/write keybinds to file

helloimhana123 committed:
Use + instead of - for keybinds

helloimhana123 committed:
Renamed struct cause it's global

helloimhana123 committed:
Show spacebar as "Spacebar" instead of "Space"

helloimhana123 committed:
Added some newlines for better help text look

helloimhana123 committed:
Use correct page when generating keybind help text

helloimhana123 committed:
Defined pattern edit keybinds

helloimhana123 committed:
Implemented pattern edit keybinds

helloimhana123 committed:
Added sample list keybinds definitions

helloimhana123 committed:
Implemented sample page keybinds

helloimhana123 committed:
Defined instrument list keybinds

helloimhana123 committed:
Check current page when updating keybind

helloimhana123 committed:
Fixed keybind released (probably)

helloimhana123 committed:
Fixed compile error

helloimhana123 committed:
Implemented instrument page keybinds

helloimhana123 committed:
Info page keybinds

helloimhana123 committed:
Keybinds order page

helloimhana123 committed:
Nav buttons should be repeatable

helloimhana123 committed:
Organized keybinds code better

helloimhana123 committed:
Fixed global keybinds not working

helloimhana123 committed:
Print keybind errors to console

helloimhana123 committed:
Palette edit page keybinds

helloimhana123 committed:
Fixed some incorrect keybind codes

helloimhana123 committed:
Fixed too long keybind text in order list

helloimhana123 committed:
Fixed pressing key twice not registering

helloimhana123 committed:
Fixed keybinds that work on multiple pages

helloimhana123 committed:
Removed shortcuts from help pages

helloimhana123 committed:
Escape menu keybinds

helloimhana123 committed:
Help page keybinds

helloimhana123 committed:
Dialog keybinds

helloimhana123 committed:
Removed text keybinds

helloimhana123 committed:
Removed duplicate keybind

helloimhana123 committed:
Removed checking specific keys in about page

helloimhana123 committed:
Waterfall keybinds

helloimhana123 committed:
Keybinds load instrument & load module

helloimhana123 committed:
Some nav keybinds

helloimhana123 committed:
Message edit keybinds

helloimhana123 committed:
Shortened dialog keybind descriptions

helloimhana123 committed:
Load sample keybinds

helloimhana123 committed:
Fixed some terrible logic for detecting key repeated

helloimhana123 committed:
Keybinds instrument list subpage

helloimhana123 committed:
Midi page keybinds

helloimhana123 committed:
Help text checks if page changed

helloimhana123 committed:
Make sure every keybind is in use

helloimhana123 committed:
Fixed keybind double press descriptions

helloimhana123 committed:
Fixed inconsistent casing in keybind descriptions

helloimhana123 committed:
Put multiple keybinds on different lines

helloimhana123 committed:
Remove keycode since it's not in use/not working

helloimhana123 committed:
Fixed some newlines in help text

helloimhana123 committed:
Update page help if page OR help text index changes

helloimhana123 committed:
Added missing pattern edit keybinds

helloimhana123 committed:
Fixed issue when loading sample would go back to open sample

helloimhana123 committed:
Fixed small issues in keybind descriptions

helloimhana123 committed:
Fixed Alt-L not selecting entire pattern

helloimhana123 committed:
Fixed all double press keybinds

helloimhana123 committed:
Fixed tested issues on pattern edit page

helloimhana123 committed:
Fixed Ctrl+Shift+N not setting track scheme

helloimhana123 committed:
Removed some unused keybinds

helloimhana123 committed:
Make sure binds are reset when on different page

helloimhana123 committed:
Fixed mod keys repeating last input

helloimhana123 committed:
Fixed global keys not working

helloimhana123 committed:
Fixed having to press twice when re-entering page

helloimhana123 committed:
Fixed panning mute channel not working

helloimhana123 committed:
Changed some key_released to key_pressed

Using key_released can  just cause a lot of issues

helloimhana123 committed:
Last of widget keybinds

helloimhana123 committed:
Fixed TODOs

helloimhana123 committed:
Fixed mess in instrument navigation

mrpapersonic commented on sys/win32/osdefs.c:
```diff
@@ -204,64 +204,77 @@ int win32_sdlevent(SDL_Event* event) {
 }
 
-void win32_create_menu(void) {
+wchar_t* str_to_wchar(char* string, int free_inputs)
```
Use `charset_iconv` or the helper macro `CHARSET_EASY_MODE`

helloimhana123 replied:
Fixed

mrpapersonic commented on schism/keybinds.c:
```diff
@@ -0,0 +1,499 @@
+#include "page.h"
+#include "dmoz.h"
+#include "keybinds_codes.c"
+#include "keybinds_init.c"
```
these should preferably be compiled separately or moved to the include folder

helloimhana123 replied:
Reason why I didn't do that is because they are not supposed to be standalone modules, they are just a way to split `keybinds.c` into different files, cause it's too long otherwise

mrpapersonic replied:
`audio_playback.c` has more lines of code itself than these three files combined. separating them into different sections within the same file, which is what a lot of the other files do as well, would work fine here.

mrpapersonic commented on include/keybinds.h:
```diff
@@ -0,0 +1,664 @@
+    int is_active;
+    enum page_numbers page;
+    int (*page_matcher)(enum page_numbers);
+} keybind_section_info;
```
these typedefs all should be suffixed with `_t` to stay consistent with the rest of the codebase (or just remove the typedef as its largely unnecessary)

helloimhana123 replied:
Fixed

mrpapersonic commented on include/keybinds.h:
```diff
@@ -0,0 +1,664 @@
+extern keybind_list global_keybinds_list;
+
+/* Key was pressed this event. Will not trigger on held down repeats. */
+#define key_pressed(SECTION, NAME) global_keybinds_list.SECTION.NAME.pressed
```
All macros should preferably be in full uppercase to indicate that it is a macro and not a function.

helloimhana123 replied:
Fixed

mrpapersonic commented on include/page.h:
```diff
@@ -386,6 +387,7 @@ enum page_numbers {
 
 	PAGE_WATERFALL,
 
+	PAGE_ANY,
```
`PAGE_GLOBAL`(?)

helloimhana123 replied:
Fixed

mrpapersonic commented on schism/keybinds.c:
```diff
@@ -0,0 +1,499 @@
+/* --------------------------------------------------------------------- */
+/* Updating bind state (handling input events) */
+
+static int check_mods(SDL_Keymod needed, SDL_Keymod current, int lenient)
```
this ought to be a macro since its just a shorthand for comparing bitfields, with the lenient version being separate

mrpapersonic commented on schism/keybinds.c:
```diff
@@ -0,0 +1,499 @@
+        } else if(sc->scancode != SDL_SCANCODE_UNKNOWN) {
+            switch(sc->scancode) {
+                case SDL_SCANCODE_RETURN:
+                    key_text = strdup("Enter");
```
[Return and Enter are two different keys.](https://daringfireball.net/2020/07/return_and_enter)

helloimhana123 replied:
There's no scancode for "enter". [scancodes](https://wiki.libsdl.org/SDL2/SDL_Scancode)

mrpapersonic replied:
Yep because it's usually only present on "extended" 101-key ANSI keyboards on the number pad. (some keyboards even mislabel the return key as an enter key as well, which is just plain wrong)

It is important to note that there really are keyboards with separate physical enter and return keys in the main section, notably old Apple portables. see the the Fn-Return issue for an example.

helloimhana123 replied:
Hmm okay so what should I do? I added this code because I wanted to help text to match how it looked before custom keybinds. There it says "Enter" and not "Return".

mrpapersonic replied:
Just change the help text to be correct then :)

mrpapersonic commented on include/util.h:
```diff
@@ -172,8 +172,33 @@ char *str_escape(const char *source, int space_hack);
+int str_count_occurences(char character, const char* str);
+
+/* count is size of array. str_array is the array. If free_inputs is true free() will be called on the strings. */
+char* str_concat_array(int count, char** str_array, int free_inputs);
```
Use variadic functions here: `str_concat(int count, int free_inputs, ...);`

mrpapersonic commented on include/util.h:
```diff
@@ -172,8 +172,33 @@ char *str_escape(const char *source, int space_hack);
+char* str_concat_with_delim(int count, char** str_array, const char* delim, int free_inputs);
+
+/* Get length of utf8 string. Utf8 strings can have a single visible character taking up 2 chars or more. */
+int utf8_length(const char* str);
```
I just added a generic charset function for this so it can be replaced with `charset_strlen(string, CHARSET_UTF8)`

helloimhana123 replied:
Fixed

mrpapersonic commented on schism/keybinds.c:
```diff
@@ -0,0 +1,499 @@
+    str_to_upper(shortcut_dup);
+
+    for (int i = 0; i < string_to_scancode_length; i++) {
+        if (strcmp(shortcut_dup, string_to_scancode[i].name) == 0) {
```
your copy of the codebase might not have this (i.e. it's too old), but `charset_strcasecmp` does exactly what you're doing here.

helloimhana123 replied:
Fixed

mrpapersonic commented on schism/util.c:
```diff
@@ -834,21 +836,171 @@ char *get_dot_directory(void)
+void str_to_upper(char *s)
+{
+    for (int i = 0; s[i]; i++)
+        s[i] = toupper((unsigned char)s[i]);
```
this sucks

helloimhana123 replied:
Oookay...? 😆 Doesn't need it anymore anyways cause of charset_strcasecmp

mrpapersonic replied:
I should've elaborated more on this lol. This would've failed on any encoding or charset that isn't what the current C locale wants which would be a lot of the time if it's being fed UTF-8.

helloimhana123 replied:
Alright gotcha

helloimhana123 committed:
Merge branch 'master' into configurable-keybinds

helloimhana123 committed:
Fix merge

helloimhana123 committed:
Use better case-insensitive string compare

helloimhana123 committed:
Use charset_strlen for utf8 string length

helloimhana123 committed:
Added _t to keybind types

helloimhana123 committed:
Changed keybind macros to use upper case

helloimhana123 committed:
Use charset_iconv instead of MS MultiByteToWideChar

helloimhana123 committed:
Renamed PAGE_ANY to PAGE_GLOBAL

helloimhana123 committed:
Use keycodes as default keybinds

helloimhana123 committed:
Notes keybinds

mrpapersonic commented on schism/keybinds_codes.c:
```diff
@@ -124,3 +124,268 @@ static struct {
+    {"BACKSPACE", SDLK_BACKSPACE},
+    {"TAB", SDLK_TAB},
+    {"SPACE", SDLK_SPACE},
+    {"EXCLAIM", SDLK_EXCLAIM},
```
These are represented as raw ASCII (e.g. `SDLK_EXCLAIM` == `'!'`) so these are a bit redundant...

helloimhana123 replied:
Sure, I just wanted to make sure I didn't miss anything so I just copied the entire list of keycodes and put it in the list.

mrpapersonic commented on schism/keybinds_codes.c:
```diff
@@ -124,3 +124,268 @@ static struct {
+    {"F21", SDLK_F21},
+    {"F22", SDLK_F22},
+    {"F23", SDLK_F23},
+    {"F23", SDLK_F24},
```
:)

helloimhana123 replied:
Same answer as above.

mrpapersonic commented on sys/win32/osdefs.c:
```diff
@@ -280,10 +281,18 @@ int win32_sdlevent(SDL_Event* event)
+	charset_error_t result = charset_iconv(string, (uint8_t**)&out, CHARSET_UTF8, CHARSET_WCHAR_T);
+
+	if (result != CHARSET_ERROR_SUCCESS) {
+		printf("Failed converting \"%s\" to wchar. Error: %i.\n", string, result);
```
`charset_iconv_error_lookup()` gives string representations.

helloimhana123 replied:
Fixed.

helloimhana123 committed:
Convert charset error to readable string

helloimhana123 committed:
Renamed "change_focus_to" to prepare for merge

helloimhana123 committed:
Merge remote-tracking branch 'upstream/master' into configurable-keybinds

helloimhana123 committed:
Added save to disk export format keybind

helloimhana123 committed:
Fixed build issues after merge

helloimhana123 committed:
Show enter key as "Return" on macos

helloimhana123 committed:
Renamed string concat functions

mrpapersonic committed:
include/charset: add missing include for non-windows

mrpapersonic committed:
keybinds_codes: guard newer keycodes behind SDL_VERSION_ATLEAST

mrpapersonic committed:
keybinds: expand indentation

keeping consistency

mrpapersonic committed:
keybinds: various styling and code improvements

mrpapersonic committed:
keybinds: add missing copyright headers

mrpapersonic commented on schism/keybinds.c:
```diff
@@ -0,0 +1,531 @@
+	int alt_needed = needed & KMOD_LALT;
+
+	int has_ralt  = current & KMOD_RALT;
+	int has_ctrl  = !has_ralt && (current & KMOD_CTRL);
```
why is this checking for `!has_ralt`?

actually, why is `KMOD_RALT` being treated differently from regular alt? they aren't separate keys. however, AltGr is, but that's mapped to `KMOD_MODE`, not `KMOD_RALT`.

mrpapersonic commented on include/keybinds.h:
```diff
@@ -0,0 +1,722 @@
+char* keybinds_get_help_text(enum page_numbers page);
+void keybinds_handle_event(struct key_event* event);
+void init_keybinds(void);
+extern keybind_list_t global_keybinds_list;
```
Having this be a globally mutable variable is kind of scary imo, but i don't see any real way around it for now

mrpapersonic committed:
keybinds: add missing brackets in switch

mrpapersonic committed:
*: rename new init methods to be prefixed with their respective file

this is in line with changes in other files as well

helloimhana123 committed:
Fixed keybinds with mods not working

helloimhana123 committed:
Support L/R mod keys (LCtrl, LAlt etc)

helloimhana123 committed:
Bad doc

helloimhana123 committed:
Merge remote-tracking branch 'upstream/master' into configurable-keybinds

mrpapersonic committed:
keybinds,util: various tweaks and stuff

keybinds_codes is now procedurally generated and uses a hash
switch statement to find the code; the keybinds themselves have
not been updated for this, however, so schism crashes on startup :)

mrpapersonic committed:
scripts/genkeybinds.py: actually use the SDL version

mrpapersonic committed:
schism/keybinds_codes.c: regen

mrpapersonic committed:
chore: merge with master

mrpapersonic committed:
page_instruments: fix malformed switch

mrpapersonic committed:
keybinds_codes: case folding!

mrpapersonic committed:
charset_unicode: whoops

mrpapersonic committed:
keybinds: optimization, free() ignores NULL pointers

mrpapersonic committed:
util: fix heap corruption bug

I apparently screwed it up, but now it's fixed (hopefully)

mrpapersonic committed:
win32/osdefs: untested crap

don't have windows nor a mingw compiler to test this with... lol

mrpapersonic committed:
config-parser: key->value to ternary operator

mrpapersonic committed:
keybinds: separate all keybinds files into separate translation units

ehhhhhh?

mrpapersonic committed:
page_info: fix global keys not working

mrpapersonic committed:
charset: whoops

mrpapersonic committed:
util: style changes

mrpapersonic committed:
keybinds: style changes and misc. optimizations

mrpapersonic committed:
keybinds_codes: reduce code clutter

mrpapersonic committed:
automake: add genkeybinds script

mrpapersonic committed:
macosx/sdlmain: add preliminary menu generation stuff

impossible to get left/right specific modifiers, so just don't
even bother...

mrpapersonic committed:
keybinds_init: str[n]dup is not portable

mrpapersonic committed:
util: str[n]dup is not portable

mrpapersonic committed:
Merge branch 'master' into configurable-keybinds

mrpapersonic committed:
util: hm

mrpapersonic committed:
Merge branch 'master' into configurable-keybinds

mrpapersonic committed:
*: properly backport time info page, fix L+R being treated as one key

mrpapersonic committed:
macosx: fix menu keybinds

oddly enough the sample and instrument library keys are warped

mrpapersonic committed:
macosx: these were wrong...

mrpapersonic committed:
osdefs/win32: this declaration was wrong

mrpapersonic committed:
automake: add keybinds header to includes

mrpapersonic committed:
CI/osx: makefile builds of utf8proc do not use rpath

mrpapersonic committed:
CI/osx: makefile builds of utf8proc do not use rpath

mrpapersonic committed:
CI/osx: ok

mrpapersonic committed:
keybinds: separate keybind init functions into their respective page units

mrpapersonic committed:
keybinds: separate keybind init functions into their respective page units

mrpapersonic committed:
keybinds: this function declaration had the wrong type

mrpapersonic committed:
keybinds: this function declaration had the wrong type

mrpapersonic committed:
keybinds_init: remove debug prints

mrpapersonic committed:
keybinds_init: remove debug prints

mrpapersonic committed:
sdlmain: undef vector, it's some powerpc altivec crap

mrpapersonic committed:
sdlmain: undef vector, it's some powerpc altivec crap

mrpapersonic committed:
CI/osx: ...

mrpapersonic committed:
fmt/iti: fix envelopes being byteswapped twice

mrpapersonic committed:
fmt/iti: fix envelopes being byteswapped twice

mrpapersonic committed:
page_timeinfo: avoid microsecond overflow

unlikely, but possible

mrpapersonic committed:
help: draw all keybinds using the Unicode font

mrpapersonic committed:
keybinds_init: fix scancode help text

mrpapersonic committed:
Merge branch 'configurable-keybinds' of ssh://github.com/HanaMcHanaface/schismtracker into configurable-keybinds

mrpapersonic committed:
Merge branch 'master' into configurable-keybinds

mrpapersonic committed:
page_waterfall: add missing keys

mrpapersonic committed:
keybinds: add (mostly) system-independent menu stuff

mrpapersonic committed:
keybinds: errr

mrpapersonic committed:
sys/macosx: remove C99 for loops

also why isn't this compiling in C99 mode? what?

mrpapersonic committed:
win32/osdefs: this was wrong, oops

mrpapersonic committed:
keybinds: check if current keybinds are not in needed

mrpapersonic committed:
keybinds: synchronize with original OS X menu

mrpapersonic committed:
win32: misc. build fixes

mrpapersonic committed:
Merge branch 'master' into configurable-keybinds

mrpapersonic committed:
Merge branch 'master' into configurable-keybinds

logiclrd commented on schism/keybinds.c:
```diff
@@ -0,0 +1,308 @@
+		}
+
+		char *old = out;
+		out = STR_CONCAT(3, old, (char*)bind->help_text, "\n");
```
This approach is inefficient and hard on the heap. Every single row results in the entire help text being copied to a newly-allocated buffer in order append that row, and then the previous one being freed. If the help text is 100 lines long, then that first line gets `strcpy`d 100 times, the second one 99 times, the third one 98 times, the fourth one 97 times, etc. Computers are fast, but this is not tight code. Also, the paradigm is copy/pasted 3 times right here. I kind of feel like it ought to be abstracted.

Since the problem here is bounded -- only the baked-in help text is ever going to be generated anyway -- my gut feeling is that it probably makes sense to just figure out what the upper bound is and do a single allocation of a buffer that'll never run out. This allocation could even just be persistent across the entire process. Some help screens will only use part of it, but that's okay.

mrpapersonic replied:
Probably a good way to go about this would be to just allocate an array of all the help texts, then use the underlying `str_concat` function to concatenate them all together, and free the array.

Also FYI - I've been incorporating many of the changes in this PR to a new `keybinds` branch in this repository, which I'm hoping will end up less abandoned :)
